test/opt/ilist_test.cpp

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include "opt/ilist.h"
+
+namespace {
+
+using ::testing::ElementsAre;
+using spvtools::utils::IntrusiveList;
+using spvtools::utils::IntrusiveNodeBase;
+using IListTest = ::testing::Test;
+
+class TestNode : public IntrusiveNodeBase<TestNode> {
+ public:
+  TestNode() : IntrusiveNodeBase<TestNode>() {}
+  int data;
+};
+
+class TestList : public IntrusiveList<TestNode> {
+ public:
+  TestList() : IntrusiveList<TestNode>() {}
+};
+
+// This test checks the push_back method, as well as using an iterator to
+// traverse the list from begin() to end().  This implicitly test the
+// PreviousNode and NextNode fucntions.
+TEST(IListTest, PushBack) {
+  TestList list;
+  TestNode nodes[10];
+  for (int i = 0; i < 10; i++) {
+    nodes[i].data = i;
+    list.push_back(&nodes[i]);
+  }
+
+  std::vector<int> output;
+  for (auto i : list) output.push_back(i.data);
+
+  EXPECT_THAT(output, ElementsAre(0, 1, 2, 3, 4, 5, 6, 7, 8, 9));
+}
+
+// Returns a list containing the values 0 to n-1 using the first n elements of
+// nodes to build the list.
+TestList BuildList(TestNode nodes[], int n) {
+  TestList list;
+  for (int i = 0; i < n; i++) {
+    nodes[i].data = i;
+    list.push_back(&nodes[i]);
+  }
+  return list;
+}
+
```

Comment:
The iterators are tested in various ways by the range-for constructs.

I'd write specific tests for begin(), end(), increment and decrement.  In particular, there's a possible ambiguity about what ++ and -- do on end and begin iterators (respectively.)

Reply:
done